sftp/public.go

```diff
@@ -0,0 +1,373 @@
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "readall fail\n")
+		return err
+	}
+	_, err = target.Write(data)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "write fail\n")
+		return err
+	}
+
+	return nil
+}
+
+// Remove removes single file under public path
+func (storage *PublishedStorage) Remove(path string) error {
+	fmt.Fprintf(os.Stderr, "Remove %s\n", path)
+	return storage.sftp.Remove(storage.expandPath(path))
+}
+
+// RemoveDirs removes directory structure under public path
+func (storage *PublishedStorage) RemoveDirs(path string, progress aptly.Progress) error {
+	fmt.Fprintf(os.Stderr, "RemoveDirs %s\n", path)
+
+	// This is non-recursive.
+	// We walk th entire tree under the prefix and delete all files.
+	// Meanwhile we put all directories into a map by their weight of slashes
+	// i.e. how deep in the tree they are.
+	dirMap := make(map[int][]string)
+	walker := storage.sftp.Walk(storage.expandPath(path))
+	for walker.Step() {
+		if err := walker.Err(); err != nil {
+			fmt.Fprintf(os.Stderr, "ERROR WHILE WALKING %s\n", err)
+			continue
+		}
+		if walker.Stat().Mode().IsDir() {
+			dir := walker.Path()
+			weight := strings.Count(dir, "/")
+			dirMap[weight] = append(dirMap[weight], dir)
+			continue
+		}
+		// If it is a file, drop it immediately.
+		err := storage.sftp.Remove(walker.Path())
+		if err != nil {
+			return err
+		}
+	}
+
+	// Now that all directories are empty all that's left is to sort the
+	// directories by their weight and then delete them in the reverse order
+	// such that the deepest directory is deleted first. Thus we always only
+	// delete empty directories.
+	var keys []int
+	for k := range dirMap {
+		keys = append(keys, k)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(keys)))
+
+	for _, k := range keys {
+		dirs := dirMap[k]
+		for _, dir := range dirs {
+			err := storage.sftp.Remove(dir)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// LinkFromPool links package file from pool to dist's pool location
+//
+// publishedDirectory is desired location in pool (like prefix/pool/component/liba/libav/)
+// sourcePool is instance of aptly.PackagePool
+// sourcePath is filepath to package file in package pool
+//
+// LinkFromPool returns relative path for the published file to be included in package index
+func (storage *PublishedStorage) LinkFromPool(publishedDirectory string,
+	sourcePool aptly.PackagePool,
+	sourcePath, sourceMD5 string, force bool) error {
+	fmt.Fprintf(os.Stderr, "LinkFromPool dir:%s pool:  srcpath:%s force:%v\n", publishedDirectory, sourcePath, force)
+
+	// verify that package pool is local pool in filesystem
+	_ = sourcePool.(*files.PackagePool)
+
+	baseName := filepath.Base(sourcePath)
+	relPath := filepath.Join(publishedDirectory, baseName)
+	poolPath := storage.expandPath(relPath)
+	fmt.Fprintf(os.Stderr, "LinkFromPool basename:%s relpath:%s poolpath:%s\n", baseName, relPath, poolPath)
+
+	// FIXME: we do not do md5 validation or anything!
```

Comment:
I've added functionality to compare the MD5 hash of a remote file in commit l1k/aptly@ffc5cbfef5d028b4a65af4079090bf6da0898cee, feel free to leverage that for your use case. As you've suggested, this could be augmented with an option to only compare metadata (timestamp, size) to achieve decent performance on a low-bandwidth connection.